### ISSUE TITLE
[Tone] High cpu usage when scrolling

### ISSUE THREAD

**Issue by LLJY** [sonyxperiadev/bug_tracker#67]
**Platform:** Tone
**Device:** (All Tone)
**Kernel version:** 4.4/5.7
**Android version:** 8.0/8.1

**Description**
cpu usage is high when scrolling

**Symptoms**
When applying a less aggresive governor profile, using userspace cpuquiet governor and set cpuquiet hal to 4min 4max cpus, scrolling still lags quite a lot.
this didn't used to happen on older builds of 8.0 and the newer kernels do not affect it. This leads me to believe that there might be a regression somewhere in the display hal.

**How to reproduce**
build latest 8.0 with the build guide.
use the cpuquiet userspace governor
set the cpuquiet hal to use 4min and 4max cpus under all scenarios.
execute this script to test this less aggresive governor profile.

`chmod 644 /sys/devices/system/cpu/cpu0/cpufreq/scaling_governor

echo interactive > /sys/devices/system/cpu/cpu0/cpufreq/scaling_governor

chmod 444 /sys/devices/system/cpu/cpu0/cpufreq/scaling_governor

chmod 644 /sys/devices/system/cpu/cpu0/cpufreq/scaling_max_freq

echo 1593600 > /sys/devices/system/cpu/cpu0/cpufreq/scaling_max_freq      

chmod 444 /sys/devices/system/cpu/cpu0/cpufreq/scaling_max_freq

chmod 644 /sys/devices/system/cpu/cpu0/cpufreq/interactive/target_loads

echo 70 480000:58 556800:75 729600:82 960000:86 1036800:2 1113600:1 1228800:96 > /sys/devices/system/cpu/cpu0/cpufreq/interactive/target_loads 

chmod 444 /sys/devices/system/cpu/cpu0/cpufreq/interactive/target_loads

chmod 644 /sys/devices/system/cpu/cpu0/cpufreq/interactive/*

echo 80000 > /sys/devices/system/cpu/cpu0/cpufreq/interactive/timer_slack

echo 307200 > /sys/devices/system/cpu/cpu0/cpufreq/interactive/hispeed_freq

echo 80000 > /sys/devices/system/cpu/cpu0/cpufreq/interactive/timer_rate

echo 0 729600:66000 960000:72000 1228800:90000 1478400:100000 > /sys/devices/system/cpu/cpu0/cpufreq/interactive/above_hispeed_delay

echo 155 > /sys/devices/system/cpu/cpu0/cpufreq/interactive/go_hispeed_load

echo 22000 > /sys/devices/system/cpu/cpu0/cpufreq/interactive/min_sample_time

echo 0 > /sys/devices/system/cpu/cpu0/cpufreq/interactive/max_freq_hysteresis

echo 0 > /sys/devices/system/cpu/cpu0/cpufreq/interactive/ignore_hispeed_on_notif

echo 0 > /sys/devices/system/cpu/cpu0/cpufreq/interactive/boost

echo 0 > /sys/devices/system/cpu/cpu0/cpufreq/interactive/fast_ramp_down

echo 0 > /sys/devices/system/cpu/cpu0/cpufreq/interactive/align_windows

echo 1 > /sys/devices/system/cpu/cpu0/cpufreq/interactive/use_migration_notif

echo 1 > /sys/devices/system/cpu/cpu0/cpufreq/interactive/use_sched_load

echo 0 > /sys/devices/system/cpu/cpu0/cpufreq/interactive/boostpulse_duration

echo 0 > /sys/devices/system/cpu/cpu0/cpufreq/interactive/io_is_busy

chmod 644 /sys/devices/system/cpu/cpu0/cpufreq/interactive/enable_prediction

echo 0 > /sys/devices/system/cpu/cpu0/cpufreq/interactive/enable_prediction

chmod 644 /sys/devices/system/cpu/cpu2/cpufreq/scaling_governor

echo interactive > /sys/devices/system/cpu/cpu2/cpufreq/scaling_governor

chmod 444 /sys/devices/system/cpu/cpu2/cpufreq/scaling_governor         

chmod 644 /sys/devices/system/cpu/cpu2/cpufreq/scaling_max_freq

echo 2265000 > /sys/devices/system/cpu/cpu2/cpufreq/scaling_max_freq            

chmod 444 /sys/devices/system/cpu/cpu2/cpufreq/scaling_max_freq

chmod 644 /sys/devices/system/cpu/cpu2/cpufreq/interactive/target_loads

echo 65 307200:69 940800:80 1036800:2 1113600:1 1248000:82 1401600:84 1824000:86 2150400:85 > /sys/devices/system/cpu/cpu2/cpufreq/interactive/target_loads 

chmod 444 /sys/devices/system/cpu/cpu2/cpufreq/interactive/target_loads

chmod 644 /sys/devices/system/cpu/cpu2/cpufreq/interactive/*

echo 90000 > /sys/devices/system/cpu/cpu2/cpufreq/interactive/timer_slack

echo 1401600 > /sys/devices/system/cpu/cpu2/cpufreq/interactive/hispeed_freq

echo 90000 > /sys/devices/system/cpu/cpu2/cpufreq/interactive/timer_rate

echo 32000 940800:75000 1248000:60000 1401600:80000 1632000:80000 > 

/sys/devices/system/cpu/cpu2/cpufreq/interactive/above_hispeed_delay

echo 400 > /sys/devices/system/cpu/cpu2/cpufreq/interactive/go_hispeed_load

echo 9000 > /sys/devices/system/cpu/cpu2/cpufreq/interactive/min_sample_time

echo 0 > /sys/devices/system/cpu/cpu2/cpufreq/interactive/max_freq_hysteresis

echo 1 > /sys/devices/system/cpu/cpu2/cpufreq/interactive/ignore_hispeed_on_notif

echo 0 > /sys/devices/system/cpu/cpu2/cpufreq/interactive/boost

echo 1 > /sys/devices/system/cpu/cpu2/cpufreq/interactive/fast_ramp_down

echo 0 > /sys/devices/system/cpu/cpu2/cpufreq/interactive/align_windows

echo 1 > /sys/devices/system/cpu/cpu2/cpufreq/interactive/use_migration_notif

echo 1 > /sys/devices/system/cpu/cpu2/cpufreq/interactive/use_sched_load

echo 0 > /sys/devices/system/cpu/cpu2/cpufreq/interactive/boostpulse_duration

echo 0 > /sys/devices/system/cpu/cpu2/cpufreq/interactive/io_is_busy

chmod 644 /sys/devices/system/cpu/cpu2/cpufreq/interactive/enable_prediction

echo 0 > /sys/devices/system/cpu/cpu2/cpufreq/interactive/enable_prediction

chmod 644 /sys/module/cpu_boost/parameters/input_boost_freq

echo 0:729600 1:0 2:940800 3:0 > /sys/module/cpu_boost/parameters/input_boost_freq

chmod 644 /sys/module/cpu_boost/parameters/input_boost_ms

echo 50 > /sys/module/cpu_boost/parameters/input_boost_ms

echo -n disable > /sys/devices/soc/soc:qcom,bcl/mode`

afterwards, go to Settings>Developer Options>Profile GPU Rendering

**Comment by jerpelea** [comment]
please retest on Android 8.1 kernel 6.4

**Comment by LLJY** [comment]
It still happens on 8.1, kernel 6.4.

**Comment by oshmoun** [comment]
is scrolling lag still an issue after updating rqbalance?

**Comment by jerpelea** [comment]
@LLJY updates?

**Comment by LLJY** [comment]
@jerpelea,  I just got EAS working on 6.4, I can confirm that the issue is still here. With EAS on 5.7, it required a small tweak to schedutil in order for it to stay at higher frequencies for longer, otherwise the scrolling lag will be unbearable.

With 6.4 and hwc2 however, EAS works with and without perfd just fine without needing any changes, however, the scrolling lag is still there, just a lot less. In comparison with Pixel (I took the powerhal and perfd+ some other blobs from it) the "profile GPU rendering" graph is significantly higher on kagura.

**Comment by jerpelea** [comment]
since  pixel and kagura are different devices there will always be diferences

if this scrolling does not make devices unusable please close the bug

**Comment by LLJY** [comment]
@jerpelea, I was just using it as an example, but the scrolling issue is still here. maybe a little less as compared to hwc1.

**Comment by kholk** [comment]
This PR helps to mitigate this issue:
https://github.com/sonyxperiadev/kernel/pull/1587

**Comment by jerpelea** [comment]
@LLJY if the mitigation is ok for you please close the bug

**Comment by oshmoun** [comment]
general GUI performance seems to have improved here on keyaki

**Comment by LLJY** [comment]
after a little bit of further testing, the issue is still here.

**Comment by jerpelea** [comment]
https://github.com/sonyxperiadev/kernel/pull/1602
more cleanup merged

**Comment by oshmoun** [comment]
This seems to be related to the composition type of the layers:
on stock, SDE is used:
```
|-----|-------------|--------|----|-------|-------------|--------------------------|---------------------|---------------------|----|------------|-----------|----|
| Idx |  Comp Type  |  Split | WB |  Pipe |    W x H    |          Format          |  Src Rect (L T R B) |  Dst Rect (L T R B) |  Z |    Flags   | Deci(HxV) | CS |
|-----|-------------|--------|----|-------|-------------|--------------------------|---------------------|---------------------|----|------------|-----------|----|
|   0 |         SDE | Comp-1 |  - | 0x080 | 1088 x 1776 |           RGBX_8888_UBWC |    0    0 1080 1776 |    0    0 1080 1776 |  0 | 0x00000000 |   0 x   0 |  0 |
|-----|-------------|--------|----|-------|-------------|--------------------------|---------------------|---------------------|----|------------|-----------|----|
|   1 |         SDE | Comp-1 |  - | 0x040 | 1088 x  144 |           RGBA_8888_UBWC |    0    0 1080  144 |    0 1776 1080 1920 |  1 | 0x00000000 |   0 x   0 |  0 |
|-----|-------------|--------|----|-------|-------------|--------------------------|---------------------|---------------------|----|------------|-----------|----|
```

While on aosp, composition is done strictly through GPU:
```
|-----|-------------|--------|----|--------|-------------|--------------------------|---------------------|---------------------|----|------------|-----------|----|-----|
| Idx |  Comp Type  |  Split | WB |  Pipe  |    W x H    |          Format          |  Src Rect (L T R B) |  Dst Rect (L T R B) |  Z |    Flags   | Deci(HxV) | CS | Rng |
|-----|-------------|--------|----|--------|-------------|--------------------------|---------------------|---------------------|----|------------|-----------|----|-----|
|   2 |  GPU_TARGET | Comp-1 |  - | 0x0020 | 1088 x 1920 |           RGBA_8888_UBWC |    0    0 1080 1920 |    0    0 1080 1920 |  0 | 0x00000000 |   0 x   0 |  0 |   0 |
|-----|-------------|--------|----|--------|-------------|--------------------------|---------------------|---------------------|----|------------|-----------|----|-----|
```

**Comment by LLJY** [comment]
its mostly been solved in 4.9, closing issue
